perfectextractor/corpora/europarl/extractor.py

```diff
@@ -114,21 +119,26 @@ def get_sentence(self, element):
         return element.xpath('ancestor::s')[0]
 
     def get_siblings(self, element, sentence_id, check_preceding):
-        path = ('preceding' if check_preceding else 'following') + '::w[ancestor::s[@id="' + sentence_id + '"]]'
-        siblings = element.xpath(path)
+        siblings = element.xpath('ancestor::s//w')
         if check_preceding:
+            siblings = siblings[:siblings.index(element)]
             siblings = siblings[::-1]
+        else:
+            siblings = siblings[siblings.index(element) + 1:]
         return siblings
 
+    def _segment_by_id(self, tree, id):
+        if tree not in self._index:
+            self._index[tree] = dict()
+            for segment in tree.xpath('//s'):
+                self._index[tree][segment.get('id')] = segment
+        return self._index[tree].get(id)
+
     def get_line_as_xml(self, tree, segment_number):
-        line = tree.xpath('//s[@id="' + segment_number + '"]')
-        if line:
-            return line[0]
-        else:
-            return None
+        return self._segment_by_id(tree, segment_number)
 
     def get_line(self, tree, segment_number):
-        line = tree.xpath('//s[@id="' + segment_number + '"]')
+        line = self._segment_by_id(tree, segment_number)
         if line:
```

Comment:
I get a warning here that `if line is not None` is preferred (like in the `get_line_and_pp` method below).